helm-v3/deisgn-proposal.md

```diff
@@ -0,0 +1,715 @@
+> 
+> [name=Adnan Abdulhussein]
+> What would the default HTTP(S) implementation be?
+> [name=Matt Fisher]
+> Adnan, would you mind clarifying your comment a little?
+> 
+> [name=Adnan Abdulhussein]
+> Sorry, the proposal mentions there will be a default implemention for `helm push` to a regular HTTP(S) chart repository. It's unclear to me how this would actually work.
+> 
+> [name=Adnan Abdulhussein]
+> Sounds like this is something that will be fleshed out more in Josh's proposal, so I'll wait for that :)
+
+## Security Considerations
+
+This architecture for Helm is more secure than Helm 2 (and equally secure to Helm Classic). The following security goals are met by this proposal:
+
+- Helm now runs with the ID and permissions of the UserAccount or ServiceAccount executing the commands.
+- RBAC can now be applied to each user
+- On-the-wire security is as secure as the cluster is configured to be (e.g. if KUBECONFIG is set up for mTLS, that's what we use)
+- Auditability is now a feature of the cluster, as each chart is "created by" the user who runs the request
+- In-cluster CRDs can be secured with RBAC, too, providing access controls for releases.
+
+
+## Appendix A: A Helm Controller
+
+This section describes a second implementation of Helm. The goal of this implementation is to provide a controller/CRD façade for the core Helm features. _However, defining the specifics of this controller was deemed outside of the scope of this proposal, which focuses on the reference implementation of the Helm CLI_.
+
+A client-only Helm follows a push-based model. To facilitate a pull-based model, we can also create a Helm controller that provides a CRD facade, but uses the Helm library.
+
+It is suggested that this (a) be a core Helm project, but (b) is kept separately from the main Helm codebase.
+
+This controller accepts Chart and values configuration as a new `HelmRequest` CRD, and then provides the following operations:
+
+- Install
+- Upgrade
+- Delete
+- Get
+- List
+
+Just like the client, the controller will operate on Release and ReleaseVersion objects. For that reason, the Helm CLI and the Helm CRD can be used in conjunction, and RBACs can even be applied to lock down certain features for one or the other. (For example, Helm CLI may be restricted to only read functions, while the controller can perform read and write operations)
+
+> [name=Adnan Abdulhussein]
+> How will the controller apply a user's RBAC permissions when installing a chart? Or is this not possible, and the controller will need to be restricted using service accounts?
+> [name=Matt Fisher]
+> It is not possible to apply a user's RBAC permissions when installing a chart via this model, which is why we decided to remove Tiller in Helm 3.
+
+This implementation will necessarily have some limitations and restrictions that the client Helm does not; namely, it will not be able to handle large charts because there is a 1M limit to CRD size.
```

Comment:
charts should come from remote servers though? just the release object should be there, and thats small. They should look like containers. you upload a small k8s object and that ends up on a node that pulls the code. the operator can work the same way. user uploads a pointer to the chart from a repo and the operator pulls it.